{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Analyse of Spatial Transcriptome data using VGP\n",
    "\n",
    "This tutorial shows loading, preprocessing, VGP analyse of Spatial Transcriptome dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages\n",
    "Here, weâ€™ll import scbean along with other popular packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scbean.model import vgp\n",
    "import pandas as pd\n",
    "import multiprocessing as mp\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading dataset\n",
    "This tutorial uses spatial transcriptome data of human breast cancer (layer 2), which contains 14,789 genes measured on 251 spots. It can be downloaded from [Spatial Transcriptomics Research](https://www.spatialresearch.org/resources-published-datasets/doi-10-1126science-aaf2403/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = 'Users/wyr/data/Layer2_BC_count_matrix-1.tsv'\n",
    "data = pd.read_csv(filepath, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GAPDH</th>\n",
       "      <th>USP4</th>\n",
       "      <th>MAPKAPK2</th>\n",
       "      <th>CPEB1</th>\n",
       "      <th>LANCL2</th>\n",
       "      <th>MCL1</th>\n",
       "      <th>TMEM109</th>\n",
       "      <th>TMEM189</th>\n",
       "      <th>ITPK1</th>\n",
       "      <th>...</th>\n",
       "      <th>TREML1</th>\n",
       "      <th>C12orf79</th>\n",
       "      <th>ZCCHC12</th>\n",
       "      <th>ZNF222</th>\n",
       "      <th>TRIM17</th>\n",
       "      <th>RNASEK</th>\n",
       "      <th>KSR2</th>\n",
       "      <th>PCDHGB4</th>\n",
       "      <th>ACOXL</th>\n",
       "      <th>CASQ2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17.907x4.967</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18.965x5.003</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18.954x5.995</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17.846x5.993</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20.016x6.019</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>23.094x23.975</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>24.981x23.964</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>21.874x24.852</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>23.096x24.93</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>24.076x25.951</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251 rows Ã— 14790 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0  GAPDH  USP4  MAPKAPK2  CPEB1  LANCL2  MCL1  TMEM109  \\\n",
       "0     17.907x4.967      1     1         1      0       0     2        0   \n",
       "1     18.965x5.003      7     0         1      0       0     0        0   \n",
       "2     18.954x5.995      5     0         0      0       0     2        0   \n",
       "3     17.846x5.993      1     0         0      0       0     2        0   \n",
       "4     20.016x6.019      2     0         1      0       0     2        0   \n",
       "..             ...    ...   ...       ...    ...     ...   ...      ...   \n",
       "246  23.094x23.975      4     0         0      0       0     1        0   \n",
       "247  24.981x23.964      4     0         0      0       0     1        0   \n",
       "248  21.874x24.852      4     0         0      0       0     0        0   \n",
       "249   23.096x24.93      2     0         0      0       0     0        1   \n",
       "250  24.076x25.951      1     0         0      0       0     0        1   \n",
       "\n",
       "     TMEM189  ITPK1  ...  TREML1  C12orf79  ZCCHC12  ZNF222  TRIM17  RNASEK  \\\n",
       "0          0      0  ...       0         0        0       0       0       0   \n",
       "1          0      0  ...       0         0        0       0       0       0   \n",
       "2          0      0  ...       0         0        0       0       0       0   \n",
       "3          0      0  ...       0         0        0       0       0       0   \n",
       "4          0      1  ...       0         0        0       0       0       0   \n",
       "..       ...    ...  ...     ...       ...      ...     ...     ...     ...   \n",
       "246        0      0  ...       0         0        0       0       0       0   \n",
       "247        0      0  ...       0         0        0       0       0       0   \n",
       "248        0      0  ...       0         0        0       0       0       0   \n",
       "249        0      0  ...       0         0        0       0       0       0   \n",
       "250        0      0  ...       0         0        0       0       0       0   \n",
       "\n",
       "     KSR2  PCDHGB4  ACOXL  CASQ2  \n",
       "0       0        0      0      0  \n",
       "1       0        0      0      0  \n",
       "2       0        0      0      0  \n",
       "3       0        0      0      0  \n",
       "4       0        0      0      0  \n",
       "..    ...      ...    ...    ...  \n",
       "246     0        0      0      0  \n",
       "247     0        0      0      0  \n",
       "248     0        0      0      0  \n",
       "249     0        0      0      0  \n",
       "250     0        0      0      0  \n",
       "\n",
       "[251 rows x 14790 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row represents a two-dimensional position and each column represents a gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing\n",
    "Here, we separate location information and gene expression from the original data to meet the input of the model. In addition, we will filter out some genes and spots with low expression levels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "location = pd.DataFrame(index=data.index)\n",
    "location['x'] = data['Unnamed: 0'].str.split('x').str.get(0).map(float)\n",
    "location['y'] = data['Unnamed: 0'].str.split('x').str.get(1).map(float)\n",
    "data.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "# Filter practically unobserved genes\n",
    "data = data.T[data.sum(0) >= 10].T\n",
    "# genes * location\n",
    "data = data.T\n",
    "location = location[data.sum(0) >= 10]\n",
    "data = data.T[data.sum(0) >= 10].T\n",
    "# model inputs\n",
    "X = location.values\n",
    "Y = data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify SV genes via VGP\n",
    "The result will be a DataFrame with p-value and adjusted p-value (q-value) for each gene.\n",
    "We identified genes with q-value less than 0.05 as spatially variable genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 9907/9907 [4:28:51<00:00,  1.63s/it]\n",
      "  0%|                                                                                 | 1/9907 [00:00<23:16,  7.09it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results....................\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 9907/9907 [00:06<00:00, 1599.96it/s]\n"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "    result = vgp.run(X, Y, mp.cpu_count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAPDH</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>USP4</td>\n",
       "      <td>0.559464</td>\n",
       "      <td>0.559514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAPKAPK2</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CPEB1</td>\n",
       "      <td>0.611947</td>\n",
       "      <td>0.611988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LANCL2</td>\n",
       "      <td>0.892471</td>\n",
       "      <td>0.897269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9902</th>\n",
       "      <td>AC245100.1</td>\n",
       "      <td>0.263375</td>\n",
       "      <td>0.263539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9903</th>\n",
       "      <td>HSD17B6</td>\n",
       "      <td>0.571247</td>\n",
       "      <td>0.571286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9904</th>\n",
       "      <td>ARNTL</td>\n",
       "      <td>0.560723</td>\n",
       "      <td>0.560871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9905</th>\n",
       "      <td>PAQR8</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9906</th>\n",
       "      <td>XRCC4</td>\n",
       "      <td>0.477873</td>\n",
       "      <td>0.477924</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9907 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene   p_value   q_value\n",
       "0          GAPDH       0.0       0.0\n",
       "1           USP4  0.559464  0.559514\n",
       "2       MAPKAPK2  0.000167  0.000167\n",
       "3          CPEB1  0.611947  0.611988\n",
       "4         LANCL2  0.892471  0.897269\n",
       "...          ...       ...       ...\n",
       "9902  AC245100.1  0.263375  0.263539\n",
       "9903     HSD17B6  0.571247  0.571286\n",
       "9904       ARNTL  0.560723  0.560871\n",
       "9905       PAQR8  0.000015  0.000015\n",
       "9906       XRCC4  0.477873  0.477924\n",
       "\n",
       "[9907 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "According to the result, we screened the SV genes, that is, the genes with q-value less than 0.05."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4129"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.sort_values([\"q_value\"], inplace=True)\n",
    "len(result[result[\"q_value\"]<0.05])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAPDH</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2857</th>\n",
       "      <td>IDH2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>693</th>\n",
       "      <td>RPL30</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2844</th>\n",
       "      <td>COL6A2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2833</th>\n",
       "      <td>METRN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2822</th>\n",
       "      <td>RAB11A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2816</th>\n",
       "      <td>DUSP14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2803</th>\n",
       "      <td>SET</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>C4A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>RPS20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene p_value q_value\n",
       "0      GAPDH     0.0     0.0\n",
       "2857    IDH2     0.0     0.0\n",
       "693    RPL30     0.0     0.0\n",
       "2844  COL6A2     0.0     0.0\n",
       "2833   METRN     0.0     0.0\n",
       "2822  RAB11A     0.0     0.0\n",
       "2816  DUSP14     0.0     0.0\n",
       "2803     SET     0.0     0.0\n",
       "706      C4A     0.0     0.0\n",
       "707    RPS20     0.0     0.0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of VGP result in R\n",
    "We mapped the expression pattern of the SV genes based on R language and some R software packages.\n",
    "\n",
    "\n",
    "**Import packages**\n",
    "\n",
    "```R\n",
    "library(ggplot2)\n",
    "library(gridExtra)\n",
    "library(grid)\n",
    "```\n",
    "\n",
    "**Load data**\n",
    "\n",
    "```R\n",
    "count <- read.csv(\"./data/Layer2_BC_expression.csv\")\n",
    "rownames(count) <- count[,1]\n",
    "count <- as.data.frame(as.matrix(count[,2:ncol(count)]))\n",
    "colnames(count) <- paste0(\"s\",1:ncol(count))\n",
    "info <- read.csv(\"./data/Layer2_BC_location.csv\")\n",
    "info <- as.data.frame(as.matrix(info[,2:ncol(info)]))\n",
    "rownames(info) <- paste0(\"s\",1:nrow(info))\n",
    "```\n",
    "\n",
    "**Define function**\n",
    "\n",
    "```R\n",
    "# anscombe variance stabilizing transformation: NB\n",
    "var_stabilize <- function(x, sv = 1) {\n",
    "  varx = apply(x, 1, var)\n",
    "  meanx = apply(x, 1, mean)\n",
    "  phi = coef(nls(varx ~ meanx + phi * meanx^2, start = list(phi = sv)))\n",
    "  return(log(x + 1/(2 * phi)))\n",
    "}\n",
    "\n",
    "# Linear normalization\n",
    "relative_func <- function(expres){\n",
    "  maxd = max(expres)-min(expres)\n",
    "  rexpr = (expres-min(expres))/maxd\n",
    "  return(rexpr)\n",
    "}\n",
    "\n",
    "# Plot gene expression pattern\n",
    "pattern_plot <- function(pltdat, igene, xy=T, main=F, title=NULL) {\n",
    "  if (!xy) {\n",
    "    xy <- matrix(as.numeric(do.call(rbind, strsplit(as.character(pltdat[,1]), split = \"x\"))), ncol = 2)\n",
    "    rownames(xy) <- as.character(pltdat[, 1])\n",
    "    colnames(xy) <- c(\"x\", \"y\")\n",
    "    pd <- cbind.data.frame(xy, pltdat[, 2:ncol(pltdat)])\n",
    "  } else {\n",
    "    pd <- pltdat\n",
    "  }\n",
    "  pal <- colorRampPalette(c(\"mediumseagreen\", \"lightyellow2\", \"deeppink\"))\n",
    "  gpt <- ggplot(pd, aes(x=x, y=y, color=pd[, igene + 2])) + geom_point(size=4) + \n",
    "    scale_color_gradientn(colours = pal(5)) + scale_x_discrete(expand=c(0, 1)) + \n",
    "    scale_y_discrete(expand=c(0, 1)) + coord_equal() + theme_bw()\n",
    "  if (main) {\n",
    "    if (is.null(title)) {\n",
    "      title = colnames(pd)[igene + 2]\n",
    "    }\n",
    "    out = gpt + labs(title=title, x=NULL, y=NULL) + \n",
    "      theme(legend.position=\"none\", plot.title=element_text(hjust=0.5, size=rel(1.5)))\n",
    "  } else {\n",
    "    out = gpt + labs(title=NULL, x=NULL, y=NULL) + theme(legend.position=\"none\")\n",
    "  }\n",
    "  return(out)\n",
    "}\n",
    "```\n",
    "\n",
    "**Plot gene expression pattern**\n",
    "\n",
    "```R\n",
    "count <- var_stabilize(count)\n",
    "pltdat  <- cbind.data.frame(info[, 1:2], apply(count[c(\"GAPDH\", \"IDH2\", \"RPL30\", \"COL6A2\", \"METRN\", \"RAB11A\", \"DUSP14\", \"SET\", \"C4A\", \"RPS20\"),], 1, relative_func))\n",
    "plot <- lapply(1:10, function(x){pattern_plot(pltdat, x, xy = T, main = T)})\n",
    "grid.arrange(grobs=plot[10:1], ncol=5)\n",
    "```\n",
    "\n",
    "<img src=\"https://github.com/jhu99/scbean/blob/vgp/docs/source/tutorials/Rplot_SV_genes.png?raw=true\" width=\"70%\">"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:tensorflow]",
   "language": "python",
   "name": "conda-env-tensorflow-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
